### PR TITLE
GPTQ int4 quantization

### PR DESCRIPTION
**Pull request by t-vi**
This adds int4 quantization, bringing the 7B model's GPU memory usage for generation to about ~5GB
The obvious next step is triton quantized matmul kernels to recover the performance.
Quantization takes <= 15 minutes when not disabling the TF32 matmuls (GPTQ upstream does this), I am not sure about the accuracy impact yet.

### CODE REVIEW

File: quantize.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        groupsize=groupsize,
+        actorder=(groupsize == -1),
+    )
+    handle = model.lm_head.register_forward_hook(gptq.collect_input_stats)
+    for j in range(inps.size(0)):
+        model.lm_head(inps[j : j + 1])
+    handle.remove()
+    q_module, error = gptq.quantize()
+    model.lm_head = q_module
+    model._original_module.lm_head = q_module  # bleh.
```
Comment (t-vi) [inline quantize.py:138]: @awaelchli Ideas for this? If I don't I get the "original" linear module.

Comment (awaelchli) [reply quantize.py:138]: I think you may want to move

`model = fabric.setup_module(model)`
after
`llama_blockwise_quantization`

Otherwise, the model is wrapped already and that looks undesired here.

Comment (awaelchli) [reply quantize.py:138]: I just opened https://github.com/Lightning-AI/lightning/issues/17285 with an idea to solve this so the user doesn't have to think about it.

Comment (lantiga) [reply quantize.py:138]: @t-vi good to merge, but do you want to try moving `model = fabric.setup_module(model)` after `llama_blockwise_quantization` as suggested before merging?

Comment (t-vi) [reply quantize.py:138]: What would be the difference to commenting out? The only step left after the quantization is to save the state dict.

Comment (lantiga) [reply quantize.py]: Probably no difference at this point. We are passing `fabric` to save_model_checkpoint` but I don't think we are leveraging any fabric-specific functionality here.

At this point I would just remove the use of fabric from `quantize.py` altogether, as it's not needed.

Comment (t-vi) [reply quantize.py]: OK

File: quantize.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+import lightning as L
+import torch
+import tqdm
+
+from lit_llama import LLaMA, Tokenizer
+from lit_llama.utils import EmptyInitOnDevice, save_model_checkpoint
+from lit_llama.quantization import GPTQQuantizer
+
+from datasets import load_dataset
```
Comment (awaelchli) [inline quantize.py:19]: we would need to add this to requirements.txt I think

File: quantize.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+from pathlib import Path
+from typing import Optional
+import gc
+
+import lightning as L
+import torch
+import tqdm
+
+from lit_llama import LLaMA, Tokenizer
+from lit_llama.utils import EmptyInitOnDevice, save_model_checkpoint
```
Comment (awaelchli) [inline quantize.py:16]:
```suggestion
from lit_llama.utils import EmptyInitOnDevice
```
unused import :)

File: quantize.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    dtype: Optional[str] = None,
+    quantize: Optional[str] = None,
+) -> None:
+    """Generates text samples based on a pre-trained LLaMA model and tokenizer.
+
+    Args:
+        datasets: The datasets to use as a comma separated string
+        # compile: Whether to compile the model.
+        accelerator: The hardware to run on. Possible choices are:
+            ``"cpu"``, ``"cuda"``, ``"mps"``, ``"gpu"``, ``"tpu"``, ``"auto"``.
```
Comment (awaelchli) [inline quantize.py:161]:
```suggestion
```

File: quantize.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    model.lm_head = q_module
+    model.lm_head.to("cpu")
+
+
+def main(
+    datasets: str = "wikitext,ptb,c4",
+    *,
+    # compilation fails as it does not support torch.complex64 for RoPE
+    # compile: bool = False,
+    accelerator: str = "auto",
```
Comment (awaelchli) [inline quantize.py:146]:
```suggestion
```

File: quantize.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    q_module, error = gptq.quantize()
+    model.lm_head = q_module
+    model.lm_head.to("cpu")
+
+
+def main(
+    datasets: str = "wikitext,ptb,c4",
+    *,
+    # compilation fails as it does not support torch.complex64 for RoPE
+    # compile: bool = False,
```
Comment (awaelchli) [inline quantize.py:145]:
```suggestion
```
I think torch.compile is not relevant for this script

File: quantize.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        t0 = time.time()
+        model = LLaMA.from_name(model_size)
+        checkpoint = torch.load(checkpoint_path)
+        model.load_state_dict(checkpoint)
+        print(f"Time to load model: {time.time() - t0:.02f} seconds.", file=sys.stderr)
+
+    model.eval()
+
+    # if compile:
+    #     model = torch.compile(model)
```
Comment (awaelchli) [inline quantize.py:209]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]
Only minor nits 🎉

File: quantize.py (hunk 2 above)
Comment (lantiga) [reply quantize.py:19]: ok let's do it

**Comment by lantiga** [comment]
Merging, let's add datasets to requirements.txt in a separate micro PR :-)

**Comment by lantiga** [comment]
Amazing contribution @t-vi !

**Comment by t-vi** [comment]
Thanks @lantiga @awaelchli !
